torch/csrc/jit/tensorexpr/kernel.cpp

```diff
@@ -1482,17 +1477,127 @@ BlockPtr TensorExprKernel::bindAllInputs() {
   return block;
 }
 
-void TensorExprKernel::compile() {
-  GRAPH_DUMP("TensorExprKernel graph:", graph_);
+void TensorExprKernel::deduceMemoryLayoutPolicy() {
+  // If the tensor is channels-last contiguous, the preferred memory layout
+  // propagation policy is to use channes-last. Otherwise, the preferred policy
+  // is to use contiguous.
+  auto _prefer_symbolic_mem =
+      [](const torch::jit::Value* val,
+         const std::vector<torch::jit::StrideInput>& stride_desc_vec) {
+        TORCH_INTERNAL_ASSERT(stride_desc_vec.size() > 0);
+        // Has symbolic stride information
+        auto cur_stride_desc = stride_desc_vec[0];
+        return (cur_stride_desc ==
+                torch::jit::StrideInput::TENSOR_CONT_CHANNELS_LAST)
+            ? MemoryLayoutPolicy::kChannelsLastNdContiguous
+            : MemoryLayoutPolicy::kContiguous;
+      };
+
+  auto _prefer_static_mem = [](const torch::jit::Value* val) {
+    // No shape info is present in the graph
+    TORCH_INTERNAL_ASSERT(
+        val->isCompleteTensor(),
+        buildErrorMessage(val->debugName() + " is not a complete tensor."));
+    const auto& tt = val->type()->expect<TensorType>();
+    const auto sizes = *tt->sizes().concrete_sizes();
+    const auto strides = *tt->strides().concrete_sizes();
+    return (c10::is_channels_last_strides_2d(sizes, strides))
+        ? MemoryLayoutPolicy::kChannelsLastNdContiguous
+        : MemoryLayoutPolicy::kContiguous;
+  };
+
+  // Filter out the tensor from the graph inputs and outputs to
+  // deduce the memory layout propagation policy
+  auto _is_tensor = [](const jit::Value* el) {
+    return el->type()->kind() == TypeKind::TensorType;
+  };
+  std::vector<torch::jit::Value*> graph_io_tensors;
+  std::copy_if(
+      graph_->inputs().begin(),
+      graph_->inputs().end(),
+      std::back_inserter(graph_io_tensors),
+      _is_tensor);
+  std::copy_if(
+      graph_->outputs().begin(),
+      graph_->outputs().end(),
+      std::back_inserter(graph_io_tensors),
+      _is_tensor);
+  // std::all_of returns true if the range is empty. But we prefer to keep
+  // the original memory layout propagation policy for this case. So we
+  // check whether the range is empty.
+  auto prefer_channels_last = (graph_io_tensors.size() > 0);
+  prefer_channels_last &= std::all_of(
+      graph_io_tensors.begin(),
+      graph_io_tensors.end(),
+      [&](const jit::Value* el) {
+        // If the tensor is neither a complete tensor nor has
+        // symbolic information, do nothing.
+        auto incomplete =
+            (!(el->isCompleteTensor())) && (!(symbolic_strides_.count(el)));
+        if (incomplete)
+          return false;
+
+        auto preferred_mem_layout = el->isCompleteTensor()
+            ? _prefer_static_mem(el)
+            : _prefer_symbolic_mem(el, symbolic_strides_[el]);
+        return preferred_mem_layout ==
+            MemoryLayoutPolicy::kChannelsLastNdContiguous;
+      });
```

Comment:
Nit: just from the code organization perspective, wouldn't it look simpler if we used a loop instead of lambdas and `std::all_of`? I think adding extra layers makes it more difficult to comprehend

Reply:
Replace the lambda and `std::all_of` with loop.